File: src/ios/CordovaStripe.swift
import Foundation

@objc(CordovaStripe) class CordovaStripe: CDVPlugin {
    
    let stripeApiUrl = "https://api.stripe.com/v1"
    
    @objc(createPaymentSession:)
    func createPaymentSession(command: CDVInvokedUrlCommand) {
        let secretKey = command.argument(at: 0) as? String ?? ""
        let mode = command.argument(at: 1) as? String ?? "payment"
        let currency = command.argument(at: 2) as? String ?? ""
        // convert amount to be int type
        let amount = command.argument(at: 3) as? Int ?? 0
        let paymentSuccessUrl = command.argument(at: 4) as? String ?? ""
        let paymentCancelUrl = command.argument(at: 5) as? String
        let priceId = command.argument(at: 6) as? String
        let customerId = command.argument(at: 7) as? String

        guard !paymentSuccessUrl.isEmpty else {
            let errorResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Please specify payment success URL")
            self.commandDelegate.send(errorResult, callbackId: command.callbackId)
            return
        }
        
        if mode.lowercased() == "subscription" {
            createSubscriptionLink(secretKey: secretKey, priceId: priceId, customerID: customerId, paymentSuccessUrl: paymentSuccessUrl, paymentCancelUrl: paymentCancelUrl, callbackId: command.callbackId)
        } else {
            createPaymentLink(secretKey: secretKey, amount: amount, currency: currency, paymentSuccessUrl: paymentSuccessUrl, paymentCancelUrl: paymentCancelUrl, callbackId: command.callbackId)
        }
    }
    
    private func createPaymentLink(secretKey: String, amount: Int, currency: String, paymentSuccessUrl: String, paymentCancelUrl: String?, callbackId: String) {
        let productName = "Stripe Product Checkout"
        
        // URL encode the success and cancel URLs
        let encodedSuccessUrl = paymentSuccessUrl.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? paymentSuccessUrl
        let encodedCancelUrl = paymentCancelUrl?.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)

        var requestBody = "line_items[0][price_data][unit_amount]=\(amount)&line_items[0][price_data][currency]=\(currency)&line_items[0][price_data][product_data][name]=\(productName)&line_items[0][quantity]=1&mode=payment&success_url=\(encodedSuccessUrl)"
        
        if let encodedCancelUrl = encodedCancelUrl {
            requestBody += "&cancel_url=\(encodedCancelUrl)"
        }

        apiCall(secretKey: secretKey, path: "/checkout/sessions", method: "POST", body: requestBody) { response, error in
            guard let response = response, let paymentLink = response["url"] as? String else {
                self.sendErrorResult(error ?? "Failed to create payment link", callbackId: callbackId)
                return
            }
            self.sendSuccessResult(paymentLink, callbackId: callbackId)
        }
    }

    
    private func createSubscriptionLink(secretKey: String, priceId: String?, customerID: String?, paymentSuccessUrl: String, paymentCancelUrl: String?, callbackId: String) {
        var requestBody = "line_items[0][price]=\(priceId ?? "")&line_items[0][quantity]=1&mode=subscription&success_url=\(paymentSuccessUrl)&customer=\(customerID ?? "")"
        if let paymentCancelUrl = paymentCancelUrl {
            requestBody += "&cancel_url=\(paymentCancelUrl)"
        }
        apiCall(secretKey: secretKey, path: "/checkout/sessions", method: "POST", body: requestBody) { response, error in
            guard let response = response, let paymentLink = response["url"] as? String else {
                self.sendErrorResult(error ?? "Failed to create subscription link", callbackId: callbackId)
                return
            }
            self.sendSuccessResult(paymentLink, callbackId: callbackId)
        }
    }
    
    private func apiCall(secretKey: String, path: String, method: String, body: String, completion: @escaping ([String: Any]?, String?) -> Void) {
        guard let url = URL(string: "\(stripeApiUrl)\(path)") else {
            completion(nil, "Invalid URL")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(secretKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = body.data(using: .utf8)
        
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            guard error == nil else {
                completion(nil, error?.localizedDescription)
                return
            }
            
            guard let data = data else {
                completion(nil, "No data received")
                return
            }

            do {
                if let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
                    if let errorJson = json["error"] as? [String: Any], let message = errorJson["message"] as? String {
                        completion(nil, message)
                    } else {
                        completion(json, nil)
                    }
                } else {
                    completion(nil, "Failed to parse JSON")
                }
            } catch {
                completion(nil, "JSON parsing error: \(error.localizedDescription)")
            }
        }
        task.resume()
    }
    
    private func sendSuccessResult(_ message: String, callbackId: String) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message)
        self.commandDelegate.send(pluginResult, callbackId: callbackId)
    }
    
    private func sendErrorResult(_ message: String, callbackId: String) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: message)
        self.commandDelegate.send(pluginResult, callbackId: callbackId)
    }
}

File: src/browser/CordovaStripe.js
// External Libraries
// Applied axios to the plugin
var script = document.createElement('script');
script.src = "https://cdn.jsdelivr.net/npm/axios/dist/axios.min.js";
document.head.appendChild(script);

// CONST
const STRIPE_API_URL = 'https://api.stripe.com/v1';

// Init Var
var browserMethod = {};

// API CALL
const apiCall = async (path, body, header, method, success, error) => {
    // Map the request method
    if(method === "POST") {
        return await axios.post(STRIPE_API_URL+path, body, {headers: header, withCredentials: true}).then(success).catch(error);
    } else {
        return await axios.get(STRIPE_API_URL+path, body, {headers: header, withCredentials: true}).then(success).catch(error);
    }
}

const convertObjectToEncoded = (object) => {
    var formBody = [];

    for (var property in object) {
        var encodedKey = encodeURIComponent(property);
        var encodedValue = encodeURIComponent(object[property]);
        formBody.push(encodedKey + "=" + encodedValue);
    }
    formBody = formBody.join("&");

    return formBody;
}

// Create a new product 
const createProduct = async (secretKey, productName) => {
    var requestBody = {
        name: productName,
    };
    requestBody = convertObjectToEncoded(requestBody);
    var requestHeader = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': 'Bearer ' + secretKey,
    }
    var responseObject = {};

    // Call Stripe API
    await apiCall('/products', requestBody, requestHeader, 'POST', 
        (data) => {
            // get product id
            responseObject.productId = data.data.id;
            responseObject.message = 'Success create product';
        } 
    , 
        (err) => {
            responseObject.productId = null;
            responseObject.message = err.data.error.message;

            // Handling error from stripe side
            if (err.response) {
                if(err.response.data.error.message) {
                    responseObject.message = err.response.data.error.message;
                } 
            }
        }
    );

    return responseObject;
}

// Create a new price 
const createPrice = async (secretKey, currency, unitAmount, productId) => {
    var requestBody = {
        unit_amount: unitAmount,
        currency: currency,
        product: productId
    };
    requestBody = convertObjectToEncoded(requestBody);
    var requestHeader = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': 'Bearer ' + secretKey,
    }
    var responseObject = {};

    // Call Stripe API
    await apiCall('/prices', requestBody, requestHeader, 'POST', 
        (data) => {
            // get price id
            responseObject.priceId = data.data.id;
            responseObject.message = "Success create price";
        } 
    , 
        (err) => {
            responseObject.priceId = null;
            responseObject.message = "Failed create price";

           // Handling error from stripe side
           if (err.response) {
            if(err.response.data.error.message) {
                responseObject.message = err.response.data.error.message;
            } 
        }
        }
    );

    return responseObject;
}

// Create a new payment link 
const createPaymentLink = async (secretKey, priceId, paymentSuccessUrl, paymentCancelUrl) => {
    var requestBody = {
        'line_items[0][price]': priceId,
        'line_items[0][quantity]': 1,
        mode: 'payment',
        success_url: paymentSuccessUrl || '',
    };
    var responseObject = {};

    if(paymentCancelUrl !== null) {
        requestBody.cancel_url = paymentCancelUrl;
    }

    requestBody = convertObjectToEncoded(requestBody);
    var requestHeader = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': 'Bearer ' + secretKey,
    }

    // Call Stripe API
    await apiCall('/checkout/sessions', requestBody, requestHeader, 'POST', 
        (data) => {
            // get payment link
            responseObject.paymentLink = data.data.url;
            responseObject.message = "Success create payment link";
        } 
    , 
        (err) => {
            responseObject.paymentLink = null;
            responseObject.message = "Failed create payment link";

            // Handling error from stripe side
            if (err.response) {
                if(err.response.data.error.message) {
                    responseObject.message = err.response.data.error.message;
                } 
            }
        }
    );

    return responseObject;
}

// Create a new subscription link 
const createSubscriptionLink = async (secretKey, priceId, customerID, itemQuantity, paymentSuccessUrl, paymentCancelUrl, error) => {
    var requestBody = {
        'line_items[0][price]': priceId,
        'line_items[0][quantity]': itemQuantity,
        mode: 'subscription',
        success_url: paymentSuccessUrl || '',
        customer: customerID
    };
    var responseObject = {};

    if(paymentCancelUrl !== null) {
        requestBody.cancel_url = paymentCancelUrl;
    }

    requestBody = convertObjectToEncoded(requestBody);
    var requestHeader = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': 'Bearer ' + secretKey,
    }

    // Call Stripe API
    await apiCall('/checkout/sessions', requestBody, requestHeader, 'POST', 
        (data) => {
            // get subscription link
            responseObject.paymentLink = data.data.url;
            responseObject.message = "Success create subscription link";
        } 
    , 
        (err) => {
            responseObject.paymentLink = null;
            responseObject.message = "Failed create subscription link";

            // Handling error from stripe side
            if (err.response) {
                if(err.response.data.error.message) {
                    responseObject.message = err.response.data.error.message;
                } 
            }
        }
    );

    return responseObject;
}

// Create new payment session
browserMethod.createPaymentSession = async (success, error, args) => {
    var secretKey = args[0];
    var mode = args[1];
    var currency = args[2];
    var amount = args[3];
    var paymentSuccessUrl = args[4];
    var paymentCancelUrl = args[5] || null;
    var priceId = args[6] || null;
    var customerID = args[7] || null;
    var itemQuantity = args[8] || null;
    var paymentLink = null;

    if(paymentSuccessUrl === null) {
        error("Please specify payment success URL");
    }

    if(mode !== 'subscription') {
        // payment mode required create test product
        // createProduct
        var product = await createProduct(secretKey, 'PayProduct');
        if(product.productId === null) {
            error(product.message);
        }

        // createPrice
        var price = await createPrice(secretKey, currency, amount, product.productId);
        if(price.priceId === null) {
            error(price.message);
        }

        // createPaymentLink
        payment = await createPaymentLink(secretKey, price.priceId, paymentSuccessUrl, paymentCancelUrl);
    } else {
        payment  = await createSubscriptionLink(secretKey, priceId, customerID, itemQuantity, paymentSuccessUrl, paymentCancelUrl);
    }


    if(payment.paymentLink === null) {
        error(payment.message);
    } else {
        success(payment.paymentLink);
    }
}

require('cordova/exec/proxy').add('CordovaStripe', browserMethod);